### PR TITLE
Network StyleGuide

### PR DESCRIPTION
**Pull request by coot**
# Description

The networking style guide based on `ouroboros-consensus/docs/StyleGuide.md`.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

File: docs/StyleGuide.md [hunk 1]
```diff
@@ -0,0 +1,1142 @@
...
+          }
+        deriving stock    (Show, Eq, Generic)
+        deriving anyclass (NoThunks, NFData)
+
+      newtype Foo = Foo {
+            unFoo :: Int
+          }
+        deriving stock   (Show)
+        deriving newtype (Eq)
+        deriving NoThunks via InspectHeapNamed "Foo" Foo
```
Comment (bolt12) [inline docs/StyleGuide.md:82]: No particular ordering on these right?

File: docs/StyleGuide.md [hunk 2]
```diff
@@ -0,0 +1,1142 @@
...
+
+      ```haskell
+      data Foo =
+          Bar Int Int
+        | Baz
+            Int
+            Int
+            (Maybe Bool)
+            [Foo]
+      ```
```
Comment (bolt12) [inline docs/StyleGuide.md:125]: I would prefer

```haskell
data Foo = Bar Int Int
         | Baz Int
               Int
               (Maybe Bool)
               [Foo]
```

If space allows, I just find it more readable with the bar aligned with the equals sign and the first argument in the same line as the constructor but not always that is possible due to space constraints

File: docs/StyleGuide.md [hunk 3]
```diff
@@ -0,0 +1,1142 @@
...
+    ```haskell
+    -- | Short title
+    --
+    -- Longer description .. 'Foo' .. "Data.String" .. @a@ .. /not/ ..
+    -- __never__ .. called \"foo bars\" .. alternative style " foo bars "
+    -- .. @'Foo' a@
+    --
+    -- > foo bar baz
+    --
+    -- .. ends here.
```
Comment (bolt12) [inline docs/StyleGuide.md:600]: Don't we usually have an empty comment line between the haddock/description and the function type signature like this:

```haskell
-- | Description ....
-- 
function :: A -> B
function = ...
```

File: docs/StyleGuide.md [hunk 4]
```diff
@@ -0,0 +1,1142 @@
...
+
+    ```haskell
+    data API m = API {
+          foo_ :: HasCallStack => Maybe a -> m a
+        }
+    foo :: HasCallStack => API m -> Maybe a -> m a
+    foo = foo_
+    ```
+
+    Without the extra wrapper `foo`, the call stack would only start at `_foo`,
```
Comment (bolt12) [inline docs/StyleGuide.md:1047]: `foo_` and not `_foo`

**Review by bolt12 — APPROVED** [review]
Looks good! I just have some small preference differences but they are that that strong of a preference to me! 

Regarding records, I prefer hanging ',' style visually, but can understand the reasoning for the other ones 😄

File: docs/StyleGuide.md (hunk 1 above)
Comment (coot) [reply docs/StyleGuide.md:82]: good question.  The order should be dictated by the type classes, not the `deriving` clause syntax, hence leaving it unspecified makes more sense to me.  What do you think?

Comment (bolt12) [reply docs/StyleGuide.md:82]: I agree!

File: docs/StyleGuide.md (hunk 2 above)
Comment (coot) [reply docs/StyleGuide.md:125]: I think this one with haddocks:

```haskell
data Foo
    = -- | Bar
      --
      Bar Int Int
      
    | -- | Baz
      --
      Baz Int
          Int
          (Maybe Bool)
          [Foo]
```

will be more consistent than one where `=` on the `data Foo =` line.

Comment (bolt12) [reply docs/StyleGuide.md:125]: Yes I agree!

File: docs/StyleGuide.md [hunk 5]
```diff
@@ -0,0 +1,1142 @@
...
+      ```haskell
+      case x of y
+        A x -> A_body
+        B y ->
+          let bl = bl_body y
+          in  B_body
+      ```
+
+      Note that we align `B_body` with `bl` in the `let` block. At the moment we
+      are not being very consistent with this.
```
[the remark this answers is not in the document]
Comment (bolt12) [reply docs/StyleGuide.md:198]: Okay!

File: docs/StyleGuide.md (hunk 3 above)
Comment (coot) [reply docs/StyleGuide.md]: Resolved.

File: docs/StyleGuide.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (coot) [reply docs/StyleGuide.md:216]: btw, the space after `in` is ok, we're just not saying it's mandatory.

**Comment by coot** [comment]
bors merge

**Pull request merged**